unixfs/archive/archive.go

```diff
@@ -36,67 +36,69 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 
 	// need to connect a writer to a reader
 	piper, pipew := io.Pipe()
+	se := func(err error) bool {
+		if err != nil {
+			pipew.CloseWithError(err)
+			return true
+		}
+		return false
+	}
 
 	// use a buffered writer to parallelize task
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
 
 	// compression determines whether to use gzip compression.
-	var maybeGzw io.WriteCloser
-	var err error
-	if compression != gzip.NoCompression {
-		maybeGzw, err = gzip.NewWriterLevel(bufw, compression)
-		if err != nil {
-			pipew.CloseWithError(err)
-			return nil, err
+	maybeGzw, err := newMaybeGzWriter(bufw, compression)
+	if se(err) {
+		return nil, err
+	}
+
+	closePipe := func() {
+		maybeGzw.Close()
+		if err := bufw.Flush(); se(err) {
+			return
 		}
-	} else {
-		maybeGzw = &identityWriteCloser{bufw}
+		pipew.Close() // everything seems to be ok.
 	}
 
 	if !archive && compression != gzip.NoCompression {
 		// the case when the node is a file
 		dagr, err := uio.NewDagReader(ctx, nd, dag)
-		if err != nil {
-			pipew.CloseWithError(err)
+		if se(err) {
```

Comment:
`se` needs to be renamed to something longer. it's so hard to see. it took me 3min to understand that `closePipe` wasn't wrong. maybe:
```
checkErrAndClosePipe
```